src/Commands/ElasticsearchIndexCommand.php

```diff
@@ -4,55 +4,255 @@
     {
-        foreach ($stores as $store) {
-            $this->indexStore($store, $indexName, $items, $dataFilter, $id);
+        foreach ($this->stores as $store) {
+            $this->indexStore(
+                store: $store,
+                indexName: $indexName,
+                items: $items,
+                id: $id,
+            );
         }
     }
 
-    public function indexStore(Store|array $store, string $indexName, callable|iterable $items, callable|array $dataFilter, callable|string $id = 'entity_id'): void
+    public function initIndex(Store|array $store, string $indexName)
     {
         $storeName = $store['name'] ?? $store['code'] ?? reset($store);
         $this->line('Indexing `' . $indexName . '` for store ' . $storeName);
+        $this->prepareIndexerWithStore($store, $indexName);
+    }
+
+    public function beforeIndex(int $totalCount)
+    {
+        if ($this->progressBar) {
+            $this->bar = $this->output->createProgressBar($totalCount);
+            $this->bar->start();
+        }
+    }
+
+    public function afterIndex()
+    {
+        $this->indexer->finish();
+        if ($this->progressBar) {
+            $this->bar->finish();
+            $this->line('');
+        }
+    }
 
+    public function indexStore(Store|array $store, string $indexName, callable|iterable|Builder $items, callable|string $id): void
+    {
+        $this->initIndex($store, $indexName);
+        $fullItems = $this->dataFrom($items);
+
+        if (is_null($fullItems)) {
+            $this->indexer->abort();
+
+            throw new Exception('Items cannot be null.');
+        }
+
+        $count = is_iterable($fullItems) ? count($fullItems) : $fullItems->getQuery()->getCountForPagination();
+
+        $this->beforeIndex($count);
+        $this->tryIndexItems($fullItems, $id);
+        $this->afterIndex();
+    }
+
+    public function tryIndexItems(iterable|Builder $items, callable|string $id): void
+    {
         try {
-            $this->prepareIndexerWithStore($store, $indexName);
-            $this->indexer->index($this->dataFrom($items), $dataFilter, $id);
-            $this->indexer->finish();
+            if (is_iterable($items)) {
+                $this->indexPartialIterable($items, $id);
+            } else {
+                $this->indexPartialQuery($items, $id);
+            }
         } catch (Exception $e) {
             $this->indexer->abort();
 
             throw $e;
         }
     }
 
-    public function prepareIndexerWithStore(Store|array $store, string $indexName, array $mapping = [], array $settings = []): void
+    public function indexPartialIterable(iterable $items, callable|string $id): void
+    {
+        if ($this->chunkSize < 1) {
+            $this->indexPartial($items, $id);
+
+            return;
+        }
+
+        $arrItems = $items instanceof Arrayable ? $items->toArray() : (array) $items;
+        foreach (array_chunk($arrItems, $this->chunkSize) as $chunk) {
```

Comment:
I feel like these chunked indexers are not the most efficient way to go about it. 
The opposite approach might be better.
Turning everything that isn't a [Generator](https://www.php.net/manual/en/class.generator.php) into a [Generator](https://www.php.net/manual/en/language.generators.syntax.php) in their most efficient way.

In case of a Builder instance, by using lazy
in case of anything paginated/chunked by looping and then advancing the pointer
In case of an array by simply looping over it or yielding it

If we need extra functionality we could wrap it in a LazyCollection https://laravel.com/docs/10.x/collections#lazy-collections